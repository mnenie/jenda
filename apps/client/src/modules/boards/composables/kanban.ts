import { computed, inject, provide, toValue } from 'vue'
import { makeDestructurable } from '@vueuse/core'
import type { InjectionKey, MaybeRefOrGetter, Ref } from 'vue'
import type { KanbanContext } from '../types/contexts'
import type { Label } from '../types'

export function useTaskCombobox<T, K extends keyof T>(items: Ref<T[]>, key: K) {
  function removeById(userId: T[K]) {
    const index = toValue(items).findIndex(u => u[key] === userId)
    if (index > -1) {
      toValue(items).splice(index, 1)
    }
  }
  function removeEntity(value: T) {
    if (value[key]) {
      removeById(value[key])
    }
  }
  function addEntity(value: T) {
    const isAdded = toValue(items).some(i => i[key] === value[key])
    isAdded ? removeById(value[key]) : toValue(items).push(value)
  }

  return makeDestructurable(
    { removeEntity, addEntity },
    [removeEntity, addEntity],
  )
}

const key: InjectionKey<KanbanContext> = Symbol('kanban')

export function provideKanbanContext<T extends KanbanContext>(value: T) {
  provide(key, value)
}

export function useKanbanContext() {
  return inject(key)!
}

export function useFilteredLabels(labels: MaybeRefOrGetter<Label[]>, search: MaybeRefOrGetter<string>) {
  return computed(() => (
    [...toValue(labels)].filter(label => label.name.toLowerCase().includes(toValue(search).toLowerCase()))
  ))
}
