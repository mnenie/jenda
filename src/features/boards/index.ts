export * from './add-board/ui';
export * from './charts/ui';
export * from './filter/ui';
