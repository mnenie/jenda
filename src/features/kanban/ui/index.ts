import AddColumn from './AddColumn.vue';
import DragCards from './DragCards.vue';
import DragColumns from './DragColumns.vue';
import SettingsSheet from './SettingsSheet.vue';
import SortingItems from './SortingItems.vue';
import RemoveBoard from './RemoveBoard.vue';

export { AddColumn, DragCards, DragColumns, SortingItems, SettingsSheet, RemoveBoard };
