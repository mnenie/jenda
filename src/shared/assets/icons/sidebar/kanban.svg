<svg xmlns="http://www.w3.org/2000/svg" width="1em" height="1em" viewBox="0 0 24 24"><g fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="1.5" color="currentColor"><path d="M2.5 12c0-4.478 0-6.718 1.391-8.109S7.521 2.5 12 2.5c4.478 0 6.718 0 8.109 1.391S21.5 7.521 21.5 12c0 4.478 0 6.718-1.391 8.109S16.479 21.5 12 21.5c-4.478 0-6.718 0-8.109-1.391S2.5 16.479 2.5 12"/><path d="M6 8c0-.943 0-1.414.293-1.707S7.057 6 8 6h.5c.943 0 1.414 0 1.707.293S10.5 7.057 10.5 8v7c0 .943 0 1.414-.293 1.707S9.443 17 8.5 17H8c-.943 0-1.414 0-1.707-.293S6 15.943 6 15zm7.5 0c0-.943 0-1.414.293-1.707S14.557 6 15.5 6h.5c.943 0 1.414 0 1.707.293S18 7.057 18 8v2c0 .943 0 1.414-.293 1.707S16.943 12 16 12h-.5c-.943 0-1.414 0-1.707-.293S13.5 10.943 13.5 10z"/></g></svg>