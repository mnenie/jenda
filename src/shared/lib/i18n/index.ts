import { createI18n } from 'vue-i18n';
import { useCookies } from '@vueuse/integrations/useCookies';
import enLocale from './locales/en-US';
import ruLocale from './locales/ru-RU';
import zhLocale from './locales/zh-CN';

type MessageSchema = typeof enLocale;

const messages = {
  'en-US': {
    ...enLocale
  },
  'ru-RU': {
    ...ruLocale
  },
  'zh-CN': {
    ...zhLocale
  }
};
const cookies = useCookies();

const getCurrentLocale = () => {
  const cookieLanguage = cookies.get('i18n');
  if (cookieLanguage) {
    return cookieLanguage;
  }

  return 'en-US';
};

const i18n = createI18n<[MessageSchema], 'en-US' | 'ru-RU' | 'zh-CN'>({
  legacy: false,
  locale: getCurrentLocale(),
  messages: messages
});

export default i18n;
