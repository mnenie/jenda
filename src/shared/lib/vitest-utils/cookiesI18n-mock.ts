import { vi } from 'vitest';

vi.mock('@vueuse/integrations/useCookies', () => {
  return {
    useCookies: () => ({
      get(key: string) {
        return key === 'i18n' ? 'en-US' : undefined;
      }
    })
  };
});
