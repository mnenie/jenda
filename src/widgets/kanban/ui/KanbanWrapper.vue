<script setup lang="ts">
import { ref } from 'vue';
import type { Column } from '@/entities/board';
import { DragCards, AddColumn, DragColumns } from '@/features/kanban';

// mocks -> after data from backend
const columns = ref<Column[]>([
  {
    _id: '0',
    title: 'TO DO',
    cards: [
      {
        _id: '12',
        title: 'Preparing onboarding materials for new UI designers',
        chat: false,
        users: [{ _id: '0', email: 'al', photoUrl: 'https://avatars.githubusercontent.com/u/121057011?v=4' }],
        priority: 'low'
      }
    ]
  },
  {
    _id: '0',
    title: 'IN PROGRESS',
    cards: [
      {
        _id: '64',
        title: 'Preparing onboarding materials for new UI designers',
        chat: true,
        users: [{ _id: '0', email: 'al', photoUrl: 'https://avatars.githubusercontent.com/u/121057011?v=4' }],
        priority: 'none'
      },
      {
        _id: '23',
        title: 'Offer notification email',
        chat: false,
        users: [
          { _id: '0', email: 'al', photoUrl: 'https://avatars.githubusercontent.com/u/121057011?v=4' },
          { _id: '1', email: 'ai', photoUrl: 'https://avatars.githubusercontent.com/u/95149637?v=4' }
        ],
        priority: 'medium'
      },
      {
        _id: '8',
        title: 'Check all internal URLs',
        tags: [
          { _id: '0', name: 'dev' },
          { _id: '1', name: 'new' }
        ],
        chat: true,
        users: [{ _id: '0', email: 'al', photoUrl: 'https://avatars.githubusercontent.com/u/95149637?v=4' }],
        priority: 'none'
      }
    ]
  }
]);
</script>

<template>
  <div :class="$style.kanban_wrapper">
    <div v-if="columns">
      <DragColumns :columns="columns">
        <template #default="{ cards }">
          <DragCards :cards="cards" />
        </template>
      </DragColumns>
    </div>
    <AddColumn />
  </div>
</template>

<style module lang="scss">
.kanban_wrapper {
  padding: 15px;
  bottom: 0;
  left: 0;
  overflow-x: auto;
  overflow-y: hidden;
  position: absolute;
  right: 0;
  scrollbar-color: rgba(var(--zinc-rgb-300), 0.4) #00000026;
  scrollbar-width: auto;
  top: 0;
  -webkit-user-select: none;
  user-select: none;
  white-space: nowrap;
  display: flex;
  flex-direction: row;
  list-style: none;
  gap: 15px;
}

.flip-list-move {
  transition: transform 0.3s;
}

.no-move {
  transition: transform 0s;
}

.ghost {
  opacity: 0.2;
}
</style>
