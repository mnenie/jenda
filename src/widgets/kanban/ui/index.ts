import KanbanWrapper from './KanbanWrapper.vue';

export { KanbanWrapper };
