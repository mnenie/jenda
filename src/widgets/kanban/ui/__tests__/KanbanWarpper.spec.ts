import { describe, it, expect } from 'vitest';
import { mount } from '@vue/test-utils';
import '@/shared/lib/vitest-utils/cookiesI18n-mock';
import KanbanWrapper from '../KanbanWrapper.vue';
import { AddColumn, DragColumns } from '@/features/kanban';
import i18n from '@/shared/lib/i18n';

describe('tests for KanbanWrapper.vue', () => {
  const wrapper = mount(KanbanWrapper, {
    global: { plugins: [i18n] },
    props: {
      columns: []
    }
  });

  it('should be render correctly', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render subcomponents', () => {
    expect(wrapper.findComponent(DragColumns).exists()).toBe(true);
    expect(wrapper.findComponent(AddColumn).exists()).toBe(true);
  });
});
