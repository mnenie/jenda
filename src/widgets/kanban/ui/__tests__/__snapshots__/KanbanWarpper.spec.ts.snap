// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tests for KanbanWrapper.vue > should be render correctly 1`] = `
"<div class="kanban_wrapper">
  <div>
    <div class="columns" tag="ul" name="flip-list" type="transition"></div>
  </div>
  <div class="text-sm block"><svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" viewBox="0 0 24 24" fill="none" stroke="var(--zinc-500)" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-plus-icon">
      <path d="M5 12h14"></path>
      <path d="M12 5v14"></path>
    </svg> Add new column</div>
  <!--teleport start-->
  <!--teleport end-->
</div>"
`;
